### PR TITLE
Make Circle CI display a useful test summary

### PR DESCRIPTION
**Pull request by islemaster**
Makes the following test suites generate JUnit XML build artifacts that Circle can parse to generate nice failure reports at the top of the test summary page (screenshot from WIP, hence the lower total test count):

screenshot from 2017-03-08 15-32-42

When tests pass it looks like this:
image

## What's included

Before we were only uploading a JUnit report for apps tests.  Here I've made the following changes:

* Add JUnit reporting when running on Circle for:
  * dashboard tests
  * pegasus tests
  * lib tests
  * shared tests
  * cucumber tests
* Move the JUnit report for apps to `$CIRCLE_TEST_REPORTS/apps` so it gets named nicely in the summary.
* pegasus and lib no longer user the `test_helper.rb` file from shared directly; instead each suite has its own `test_helper.rb` file, and they each require a `common_test_helper.rb` that lives in shared - this lets us do some independent reporting setup in each suite that keeps them from messing with each others' settings.

## What's _not_ working

There's also [a known issue](https://github.com/code-dot-org/code-dot-org/pull/10907#issuecomment-254838470) with cucumber test failures not reporting, because some of the reports don't get uploaded as build artifacts; I think this is a bug in how we run Cucumber in parallel, and not with actually uploading the artifacts.  I'm [testing for this on a separate branch](https://circleci.com/gh/code-dot-org/code-dot-org/tree/minitest-junit-reporting-with-failure) and you can see the problem [demonstrated in this build](https://circleci.com/gh/code-dot-org/code-dot-org/32668).  If this does happen the _build_ will still report failure and we can track the problem down in the logs as we do now - it just won't be surfaced in the summary at the top of the build page.  I propose that this is still worth doing as it's significantly better than the complete lack of summary that we have right now.

### CODE REVIEW

File: pegasus/test/test_pdf_merger.rb [hunk 1]
```diff
@@ -29,57 +30,89 @@ def setup
...
+      }
+    ].each do |dependency|
+      found = system 'which', dependency[:name], out: '/dev/null'
+      assert found, <<-MSG
+Expected '#{dependency[:name]}' to be installed.
+      This test depends on #{dependency[:name]} (#{dependency[:url]})
+      You should install it locally:
+         OSX: #{dependency[:osx]}
+      Ubuntu: #{dependency[:ubuntu]}
+      MSG
```
Comment (aoby) [inline pegasus/test/test_pdf_merger.rb]: nice :)

File: pegasus/test/test_pdf_merger.rb [hunk 2]
```diff
@@ -29,57 +30,89 @@ def setup
...
+      {
+        name: 'gs',
+        url: 'https://www.ghostscript.com/',
+        osx: 'brew install gs',
+        ubuntu: 'sudo apt-get install ghostscript'
+      },
+      {
+        name: 'pdftk',
+        url: 'https://www.pdflabs.com/tools/pdftk-the-pdf-toolkit/',
+        osx: 'brew install https://raw.github.com/quantiverge/homebrew-binary/pdftk/pdftk.rb',
```
Comment (davidsbailey) [inline pegasus/test/test_pdf_merger.rb:43]: This command isn't working for OS X as alluded to by the instructions in SETUP.md. I think this is probably fine for now and if we fix the setup instructions we can try to remember to fix it here too.

**Comment by davidsbailey** [comment]
OSX failures aside, I'm getting a different error when I try to run only the pdf merger test. Do you repro?
```
Davids-MacBook-Pro-2:~/src/cdo2/pegasus (minitest-junit-reporting)$ ruby -Itest ./test/test_pdf_merger.rb 
/Users/dsb/.rbenv/versions/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:121:in `require': cannot load such file -- cdo/db (LoadError)
	from /Users/dsb/.rbenv/versions/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:121:in `require'
	from /Users/dsb/src/cdo2/shared/test/common_test_helper.rb:10:in `<top (required)>'
	from /Users/dsb/src/cdo2/pegasus/test/test_helper.rb:2:in `require_relative'
	from /Users/dsb/src/cdo2/pegasus/test/test_helper.rb:2:in `<top (required)>'
	from ./test/test_pdf_merger.rb:4:in `require_relative'
	from ./test/test_pdf_merger.rb:4:in `<main>'
```

**Comment by islemaster** [comment]
Updated test summary after making sure shared/pegasus/lib were treated separately:

> Your build ran 3004 tests in apps, dashboard, lib, pegasus, shared, Cucumber with 0 failures
> Slowest test: Basic navigation for PLC stuff Basic navigation and ribbon changing works as expected (took 294.32 seconds).

**Comment by islemaster** [comment]
Whoops!  @davidsbailey that last error you reported is fixed in 59a043f - I had some order-dependent requires mixed up.

**Comment by islemaster** [comment]
Would one of you mind pulling this branch and verifying that the test_pdf_merger.rb tests are skipped if pdftk isn't found?

**Comment by davidsbailey** [comment]
Without pdftk, the individual test results say they are errors, but the summary says they are skips:
```
PDFMergerTest
Expected 'pdftk' to be installed.
    This test depends on pdftk (https://www.pdflabs.com/tools/pdftk-the-pdf-toolkit/)
    You should install it locally:
       OSX: brew install https://raw.github.com/quantiverge/homebrew-binary/pdftk/pdftk.rb
    Ubuntu: sudo apt-get install pdftk
  test_merge_two_remote_pdfs                                     ERROR (0.01s)
... (5 more similar ERRORs)
Finished in 0.04060s
6 tests, 0 assertions, 0 failures, 0 errors, 6 skips
```

**Comment by aoby** [comment]
I've been seeing this in action recently, and I love it. It's saved me a lot of time searching for errors in Circle logs! 😄

**Pull request merged**